File: lib/core/project_manager.py
import re
import shutil
from pathlib import Path
from typing import Any, Dict
import yaml

from core.config_adapter import wrap_config, ConfigDict

ROOT_DIR = Path(__file__).parent.parent.parent.resolve()
DEFAULT_CONFIG_PATH = ROOT_DIR / "config.yaml"


class ProjectPaths:
    def __init__(
        self,
        project_name: str,
        project_dir: Path,
        src_dir: Path,
        workspace_dir: Path,
        output_dir: Path,
        config_path: Path
    ):
        self.project_name = project_name
        self.project_dir = project_dir
        self.src_dir = src_dir
        self.workspace_dir = workspace_dir
        self.output_dir = output_dir
        self.config_path = config_path

    def load_config(self) -> ConfigDict:
        """Loads project-specific config.yaml, deep-overlaying onto root default config."""
        config = ConfigDict()
        if DEFAULT_CONFIG_PATH.exists():
            with open(DEFAULT_CONFIG_PATH, "r", encoding="utf-8") as f:
                root_dict = yaml.safe_load(f) or {}
                config.deep_merge(root_dict)

        if self.config_path.exists():
            with open(self.config_path, "r", encoding="utf-8") as f:
                proj_dict = yaml.safe_load(f) or {}
                config.deep_merge(proj_dict)

        # Ensure project-specific workspace_dir and output_dir are set in config dict
        config["workspace_dir"] = str(self.workspace_dir)
        config["output_dir"] = str(self.output_dir)
        return config


class ProjectManager:
    @staticmethod
    def resolve_project_paths(target_path: Path | str) -> ProjectPaths:
        """
        Resolves project layout (src, workspace, output, config.yaml) for a given file or directory target.
        - Assets targets like 'assets/foods/src/v1.mp4' -> project 'foods' in 'assets/foods/'
        - Targets like 'input/foods/links.txt' -> project 'foods' in 'assets/foods/'
        - Independent targets -> project 'default' in 'assets/default/'
        """
        target_path = Path(target_path).resolve()
        assets_dir = ROOT_DIR / "assets"

        # Check if target is inside assets/<project_name>/...
        project_name = "default"
        try:
            rel_to_assets = target_path.relative_to(assets_dir)
            parts = rel_to_assets.parts
            if parts:
                project_name = parts[0]
        except ValueError:
            # Check legacy input/<project_name>/...
            try:
                legacy_input = ROOT_DIR / "input"
                rel_to_input = target_path.relative_to(legacy_input)
                parts = rel_to_input.parts
                if parts and parts[0] not in ("src", "downloaded"):
                    project_name = parts[0]
            except ValueError:
                # If target_path is a directory inside assets or named project
                if target_path.parent == assets_dir:
                    project_name = target_path.name

        project_dir = assets_dir / project_name
        src_dir = project_dir / "src"
        workspace_dir = project_dir / "workspace"
        output_dir = project_dir / "output"
        config_path = project_dir / "config.yaml"

        # Ensure project subdirectories exist
        src_dir.mkdir(parents=True, exist_ok=True)
        workspace_dir.mkdir(parents=True, exist_ok=True)
        output_dir.mkdir(parents=True, exist_ok=True)

        # Copy default root config.yaml if project config does not exist
        if not config_path.exists() and DEFAULT_CONFIG_PATH.exists():
            shutil.copy(str(DEFAULT_CONFIG_PATH), str(config_path))

        return ProjectPaths(
            project_name=project_name,
            project_dir=project_dir,
            src_dir=src_dir,
            workspace_dir=workspace_dir,
            output_dir=output_dir,
            config_path=config_path
        )

    @staticmethod
    def get_job_id(video_path: Path | str) -> str:
        """
        Generates a clean, deterministic job ID based on the video filename.
        Example: 'video_001.mp4' -> 'job_video_001'
        """
        stem = Path(video_path).stem
        sanitized = re.sub(r"[^a-zA-Z0-9_\-]", "_", stem)
        return f"job_{sanitized}"

File: video_engine/lib/core/project_manager.dart
import 'dart:io';
import 'package:path/path.dart' as p;
import 'config_adapter.dart';

class ProjectPaths {
  final String projectName;
  final Directory projectDir;
  final Directory srcDir;
  final Directory cutDir;
  final Directory mergeDir;
  final Directory workspaceDir;
  final Directory outputDir;
  final File configPath;
  final File rootConfigPath;

  const ProjectPaths({
    required this.projectName,
    required this.projectDir,
    required this.srcDir,
    required this.cutDir,
    required this.mergeDir,
    required this.workspaceDir,
    required this.outputDir,
    required this.configPath,
    required this.rootConfigPath,
  });

  /// Loads configuration merged with root config.yaml fallback
  ConfigDict loadConfig([Map<String, dynamic>? overrides]) {
    ConfigDict config;
    if (configPath.existsSync()) {
      config = ConfigDict.fromYamlFile(configPath);
    } else if (rootConfigPath.existsSync()) {
      config = ConfigDict.fromYamlFile(rootConfigPath);
    } else {
      config = ConfigDict(<String, dynamic>{});
    }

    if (overrides != null && overrides.isNotEmpty) {
      config.deepMerge(overrides);
    }

    // Set standard runtime path overrides
    config['workspace_dir'] = workspaceDir.path;
    config['output_dir'] = outputDir.path;
    config['cut_dir'] = cutDir.path;
    config['merge_dir'] = mergeDir.path;

    return config;
  }
}

class ProjectManager {
  /// Robust root directory finder that works in macOS GUI sandboxes, CLI, and dev environments.
  static Directory findRootDir([String? startPath]) {
    Directory? curr;
    if (startPath != null && startPath.isNotEmpty) {
      final f = File(startPath);
      curr = f.existsSync() ? f.parent.absolute : Directory(startPath).absolute;
    } else {
      final envRoot = Platform.environment['SUB_VIDEO_ROOT'];
      if (envRoot != null && Directory(envRoot).existsSync()) {
        return Directory(envRoot).absolute;
      }
      curr = Directory.current.absolute;
    }

    // 1. Walk upwards looking for repository markers
    Directory checkDir = curr;
    for (int i = 0; i < 10; i++) {
      final hasResources = Directory(p.join(checkDir.path, 'resources')).existsSync() ||
          Directory(p.join(checkDir.path, 'assets')).existsSync();
      final hasConfig = File(p.join(checkDir.path, 'config.yaml')).existsSync();
      final hasPubspec = File(p.join(checkDir.path, 'pubspec.yaml')).existsSync();
      final hasVideoEngine = Directory(p.join(checkDir.path, 'video_engine')).existsSync();

      if (hasResources && (hasConfig || hasPubspec || hasVideoEngine)) {
        return checkDir;
      }

      final parent = checkDir.parent;
      if (parent.path == checkDir.path) break;
      checkDir = parent;
    }

    // 2. Standard Sub-Video workspace location fallback
    final standardHome = Platform.environment['HOME'];
    if (standardHome != null) {
      final standardWorkspace = Directory(p.join(standardHome, 'Documents', 'projects', 'video', 'Sub-Video'));
      if (standardWorkspace.existsSync()) {
        return standardWorkspace;
      }
    }

    return Directory.current.absolute;
  }

  /// Resolves all paths for a project given a project directory or path inside the project.
  static ProjectPaths resolveProjectPaths(String targetPath, {Directory? rootDirOverride}) {
    final rootDir = rootDirOverride ?? findRootDir(targetPath);
    final target = File(targetPath).existsSync() ? File(targetPath).parent : Directory(targetPath);

    String projectName;
    Directory projectDir;

    if (target.path.contains('/resources/') || target.path.contains('\\resources\\')) {
      final parts = p.split(p.normalize(target.absolute.path));
      final resIndex = parts.lastIndexOf('resources');
      if (resIndex != -1 && resIndex + 1 < parts.length) {
        projectName = parts[resIndex + 1];
        projectDir = Directory(p.join(rootDir.path, 'resources', projectName));
      } else {
        projectName = p.basename(target.path);
        projectDir = target;
      }
    } else if (target.path.contains('/assets/') || target.path.contains('\\assets\\')) {
      final parts = p.split(p.normalize(target.absolute.path));
      final assetsIndex = parts.lastIndexOf('assets');
      if (assetsIndex != -1 && assetsIndex + 1 < parts.length) {
        projectName = parts[assetsIndex + 1];
        projectDir = Directory(p.join(rootDir.path, 'resources', projectName)).existsSync()
            ? Directory(p.join(rootDir.path, 'resources', projectName))
            : Directory(p.join(rootDir.path, 'assets', projectName));
      } else {
        projectName = p.basename(target.path);
        projectDir = target;
      }
    } else if (p.basename(target.parent.path) == 'resources' || p.basename(target.parent.path) == 'assets') {
      projectName = p.basename(target.path);
      projectDir = target;
    } else {
      projectName = p.basename(target.path);
      final resCandidate = Directory(p.join(rootDir.path, 'resources', projectName));
      final assetCandidate = Directory(p.join(rootDir.path, 'assets', projectName));
      if (resCandidate.existsSync()) {
        projectDir = resCandidate;
      } else if (assetCandidate.existsSync()) {
        projectDir = assetCandidate;
      } else {
        projectDir = target;
      }
    }

    final defaultRootConfig = File(p.join(rootDir.path, 'config.yaml'));
    final srcDir = Directory(p.join(projectDir.path, 'src'));
    final cutDir = Directory(p.join(projectDir.path, 'cut'));
    final mergeDir = Directory(p.join(projectDir.path, 'merge'));
    final workspaceDir = Directory(p.join(projectDir.path, 'workspace'));
    final outputDir = Directory(p.join(projectDir.path, 'output'));
    final configPath = File(p.join(projectDir.path, 'config.yaml'));

    srcDir.createSync(recursive: true);
    cutDir.createSync(recursive: true);
    mergeDir.createSync(recursive: true);
    workspaceDir.createSync(recursive: true);
    outputDir.createSync(recursive: true);

    // Copy default root config if project config does not exist
    if (!configPath.existsSync() && defaultRootConfig.existsSync()) {
      defaultRootConfig.copySync(configPath.path);
    }

    return ProjectPaths(
      projectName: projectName,
      projectDir: projectDir,
      srcDir: srcDir,
      cutDir: cutDir,
      mergeDir: mergeDir,
      workspaceDir: workspaceDir,
      outputDir: outputDir,
      configPath: configPath,
      rootConfigPath: defaultRootConfig,
    );
  }

  /// Extracts the base stem of a video, stripping common output suffixes like _vi, _narrated, _translated.
  static String extractCleanStem(String videoPath) {
    var stem = p.basenameWithoutExtension(videoPath);
    stem = stem.replaceAll(RegExp(r'(_vi|_narrated|_translated|_trans|_sub|_voice)$', caseSensitive: false), '');
    return stem;
  }

  /// Generates a clean, deterministic job ID based on the video filename.
  /// Example: 'video_001.mp4' -> 'job_video_001', 'output/video_001_vi.mp4' -> 'job_video_001'
  static String getJobId(String videoPath) {
    final stem = extractCleanStem(videoPath);
    final sanitized = stem.replaceAll(RegExp(r'[^a-zA-Z0-9_\-]'), '_');
    return 'job_$sanitized';
  }

  /// Resolves the true source video file from either src/ or output/ paths.
  static File resolveSourceVideo(String videoPath, {Directory? projectDir}) {
    final file = File(videoPath);
    if (!file.existsSync()) return file;

    final parentDir = file.parent.path;
    final isOutput = parentDir.endsWith('/output') || parentDir.endsWith('\\output') || p.basename(parentDir) == 'output';

    if (isOutput) {
      final cleanStem = extractCleanStem(videoPath);
      final proj = projectDir ?? file.parent.parent;
      final srcDir = Directory(p.join(proj.path, 'src'));
      if (srcDir.existsSync()) {
        final supportedExtensions = ['.mp4', '.mov', '.mkv', '.avi', '.webm', '.flv', '.ts'];
        for (final ext in supportedExtensions) {
          final candidate = File(p.join(srcDir.path, '$cleanStem$ext'));
          if (candidate.existsSync()) {
            return candidate;
          }
        }
      }
    }

    return file;
  }
}
